File: ohara-client/src/main/scala/com/island/ohara/client/kafka/TopicAdmin.scala
/*
 * Copyright 2019 is-land
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.island.ohara.client.kafka

import java.util
import java.util.concurrent.ExecutionException
import java.util.concurrent.atomic.AtomicBoolean
import java.util.{Collections, Objects, Properties}

import com.island.ohara.common.annotations.Optional
import com.island.ohara.common.setting.TopicKey
import com.island.ohara.common.util.{CommonUtils, Releasable}
import org.apache.kafka.clients.CommonClientConfigs
import org.apache.kafka.clients.admin.{AdminClient, NewPartitions, NewTopic, TopicDescription}
import org.apache.kafka.clients.consumer.{ConsumerConfig, KafkaConsumer}
import org.apache.kafka.common.TopicPartition
import org.apache.kafka.common.config.{ConfigResource, TopicConfig}
import org.apache.kafka.common.serialization.ByteArrayDeserializer

import scala.collection.JavaConverters._
import scala.concurrent.{Future, Promise}

/**
  * this is a wrap of kafka's AdminClient. However, we only wrap the functions about "topic" since the others are useless
  * to us.
  */
trait TopicAdmin extends Releasable {
  /**
    * Change the number of partitions of topic.
    * Currently, reducing the number of partitions is not allowed!
    * @param topicKey topic key
    * @param numberOfPartitions the partitions that given topic should have
    * @return topic information
    */
  def changePartitions(topicKey: TopicKey, numberOfPartitions: Int): Future[Unit]

  /**
    * list all topics
    * @return topics information
    */
  def topics(): Future[Seq[TopicAdmin.KafkaTopicInfo]]

  /**
    * check the existence of topic on remote broker cluster
    * @param topicKey topic key
    * @return true if the topic lies in the cluster. Otherwise, false
    */
  def exist(topicKey: TopicKey): Future[Boolean]

  /**
    * start a process to create topic
    * @return topic creator
    */
  def creator: TopicAdmin.Creator

  /**
    * delete a existent topic
    * @param topicKey topic key
    * @return true if it does delete a topic. otherwise, false
    */
  def delete(topicKey: TopicKey): Future[Boolean] = delete(topicKey.topicNameOnKafka())

  def delete(topicName: String): Future[Boolean]

  /**
    * the connection information to kafka's broker
    * @return connection props
    */
  def connectionProps: String

  def closed: Boolean
}

object TopicAdmin {
  def apply(_connectionProps: String): TopicAdmin = new TopicAdmin {
    private[this] val _closed            = new AtomicBoolean(false)
    override val connectionProps: String = _connectionProps

    override def closed: Boolean = _closed.get()

    /**
      * extract the exception wrapped in ExecutionException.
      * @param f action
      * @tparam T return type
      * @return return value
      */
    private[this] def unwrap[T](f: () => T): T =
      try f()
      catch {
        case e: ExecutionException =>
          throw e.getCause
      }
    private[this] def toAdminProps(connectionProps: String): Properties = {
      val adminProps = new Properties
      adminProps.put(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, Objects.requireNonNull(connectionProps))
      adminProps
    }

    private[this] val admin = AdminClient.create(toAdminProps(connectionProps))

    override def close(): Unit = if (_closed.compareAndSet(false, true)) Releasable.close(admin)

    override def creator: Creator =
      (topicKey: TopicKey, numberOfPartitions: Int, numberOfReplications: Short, configs: Map[String, String]) => {
        val promise = Promise[Unit]
        unwrap(
          () =>
            admin
              .createTopics(
                util.Collections.singletonList(
                  new NewTopic(topicKey.topicNameOnKafka(), numberOfPartitions, numberOfReplications)
                    .configs(configs.asJava)
                )
              )
              .values()
              .get(topicKey.topicNameOnKafka())
              .whenComplete((_, exception) => {
                if (exception == null)
                  promise.success(Unit)
                else promise.failure(exception)
              })
        )
        promise.future
      }
    type TopicName       = String
    type Partition       = Int
    type BeginningOffset = Long
    type EndOffset       = Long

    private[this] def toTopicInfo(
      topicDescription: TopicDescription,
      configs: Map[String, String],
      offsets: Map[TopicName, Map[Partition, (BeginningOffset, EndOffset)]]
    ): KafkaTopicInfo = {
      new KafkaTopicInfo(
        name = topicDescription.name(),
        numberOfPartitions = topicDescription.partitions().size(),
        numberOfReplications = topicDescription.partitions().get(0).replicas().size().asInstanceOf[Short],
        partitionInfos = topicDescription.partitions().asScala.map { kafkaPartition =>
          val (beginningOffset, endOffset) =
            offsets.get(topicDescription.name()).flatMap(ps => ps.get(kafkaPartition.partition())).getOrElse((-1L, -1L))
          new KafkaPartitionInfo(
            index = kafkaPartition.partition(),
            leaderNode = kafkaPartition.leader().host(),
            replicaNodes = kafkaPartition.replicas().asScala.map(_.host()).toSet,
            inSyncReplicaNodes = kafkaPartition.isr().asScala.map(_.host()).toSet,
            beginningOffset = beginningOffset,
            endOffset = endOffset
          )
        },
        configs = configs
      )
    }

    /**
      * fetch the low/high offsets from all topics.
      * the client code is KafkaConsumer than KafkaAdmin since the later is unsupported to fetch offsets ...
      */
    private[this] def topicOffsets(): Map[TopicName, Map[Partition, (BeginningOffset, EndOffset)]] = {
      val config = new Properties
      config.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, _connectionProps)
      config.setProperty(ConsumerConfig.CLIENT_ID_CONFIG, CommonUtils.randomString())
      val consumer =
        new KafkaConsumer[Array[Byte], Array[Byte]](config, new ByteArrayDeserializer, new ByteArrayDeserializer)
      try {
        consumer
          .listTopics()
          .asScala
          .map {
            case (topicName, pts) => topicName -> pts.asScala.map(p => new TopicPartition(p.topic(), p.partition()))
          }
          .map {
            case (topicName, pts) =>
              def fetchOffset(offsets: Map[TopicPartition, java.lang.Long], partition: TopicPartition): Long =
                offsets.getOrElse(
                  partition,
                  throw new NoSuchElementException(
                    s"topic:${partition.topic()} partition:${partition.partition()} does not exist"
                  )
                )
              val beginningOffsets = consumer.beginningOffsets(pts.asJava).asScala.toMap
              val endOffsets       = consumer.endOffsets(pts.asJava).asScala.toMap
              topicName -> pts
                .map(pt => pt.partition() -> (fetchOffset(beginningOffsets, pt) -> fetchOffset(endOffsets, pt)))
                .toMap
          }
          .toMap
      } finally Releasable.close(consumer)
    }

    override def topics(): Future[Seq[KafkaTopicInfo]] =
      try {
        val allOffsets = topicOffsets()
        val promise    = Promise[Seq[KafkaTopicInfo]]
        unwrap(
          () =>
            admin
              .listTopics()
              .names()
              .whenComplete((topicNames, exception) => {
                if (exception == null)
                  admin
                    .describeTopics(topicNames)
                    .all()
                    .whenComplete((topicDescriptions, exception) => {
                      if (exception == null)
                        admin
                          .describeConfigs(
                            topicDescriptions
                              .keySet()
                              .asScala
                              .map(name => new ConfigResource(ConfigResource.Type.TOPIC, name))
                              .asJava
                          )
                          .all()
                          .whenComplete(
                            (topicsAndConfigs, exception) =>
                              try {
                                if (exception == null)
                                  promise.success(
                                    topicDescriptions
                                      .values()
                                      .asScala
                                      .map { topicDescription =>
                                        val configs = topicsAndConfigs.asScala
                                          .find(_._1.name() == topicDescription.name())
                                          .map(_._2.entries().asScala.map(e => e.name() -> e.value()).toMap)
                                          .getOrElse(Map.empty)
                                        toTopicInfo(topicDescription, configs, allOffsets)
                                      }
                                      .toSeq
                                  )
                                else promise.failure(exception)
                              } catch {
                                case e: Throwable => promise.failure(e)
                              }
                          )
                      else promise.failure(exception)
                    })
                else promise.failure(exception)
              })
        )
        promise.future
      } catch {
        case e: Throwable => Future.failed(e)
      }

    override def changePartitions(topicKey: TopicKey, numberOfPartitions: Int): Future[Unit] = {
      val promise = Promise[Unit]
      unwrap(
        () =>
          admin
            .createPartitions(
              Collections.singletonMap(topicKey.topicNameOnKafka(), NewPartitions.increaseTo(numberOfPartitions))
            )
            .all()
            .whenComplete((_, exception) => {
              if (exception == null) promise.success(Unit)
              else promise.failure(exception)
            })
      )
      promise.future
    }

    override def delete(topicName: String): Future[Boolean] = {
      val promise = Promise[Boolean]
      unwrap(
        () =>
          admin
            .listTopics()
            .names()
            .whenComplete((topicNames, exception) => {
              if (exception == null) {
                if (topicNames.asScala.contains(topicName))
                  admin
                    .deleteTopics(util.Collections.singletonList(topicName))
                    .all()
                    .whenComplete((_, exception) => {
                      if (exception == null) promise.success(true)
                      else promise.failure(exception)
                    })
                else
                  promise.success(false)
              } else promise.failure(exception)
            })
      )
      promise.future
    }

    override def exist(topicKey: TopicKey): Future[Boolean] = {
      val promise = Promise[Boolean]
      unwrap(
        () =>
          admin
            .listTopics()
            .names()
            .whenComplete((topicNames, exception) => {
              if (exception == null) promise.success(topicNames.asScala.contains(topicKey.topicNameOnKafka()))
              else promise.failure(exception)
            })
      )
      promise.future
    }
  }

  final class KafkaPartitionInfo(
    val index: Int,
    val leaderNode: String,
    val replicaNodes: Set[String],
    val inSyncReplicaNodes: Set[String],
    val beginningOffset: Long,
    val endOffset: Long
  )

  final class KafkaTopicInfo(
    val name: String,
    val numberOfPartitions: Int,
    val numberOfReplications: Short,
    val partitionInfos: Seq[KafkaPartitionInfo],
    val configs: Map[String, String]
  )

  trait Creator extends com.island.ohara.common.pattern.Creator[Future[Unit]] {
    private[this] var topicKey: TopicKey          = _
    private[this] var numberOfPartitions: Int     = 1
    private[this] var numberOfReplications: Short = 1
    private[this] var configs: Map[String, String] = Map(
      TopicConfig.CLEANUP_POLICY_CONFIG -> CleanupPolicy.DELETE.name
    )
    def topicKey(topicKey: TopicKey): Creator.this.type = {
      this.topicKey = Objects.requireNonNull(topicKey)
      this
    }

    @Optional("default is 1")
    def numberOfPartitions(numberOfPartitions: Int): Creator = {
      this.numberOfPartitions = CommonUtils.requirePositiveInt(numberOfPartitions)
      this
    }

    @Optional("default is 1")
    def numberOfReplications(numberOfReplications: Short): Creator = {
      this.numberOfReplications = CommonUtils.requirePositiveShort(numberOfReplications)
      this
    }

    @Optional("default is CleanupPolicy.DELETE")
    def cleanupPolicy(cleanupPolicy: CleanupPolicy): Creator = {
      this.configs ++= Map(TopicConfig.CLEANUP_POLICY_CONFIG -> cleanupPolicy.name)
      this
    }

    def config(key: String, value: String): Creator = {
      this.configs ++= Map(CommonUtils.requireNonEmpty(key) -> CommonUtils.requireNonEmpty(value))
      this
    }

    def configs(configs: Map[String, String]): Creator = {
      this.configs ++= Objects.requireNonNull(configs)
      this
    }

    override def create(): Future[Unit] = doCreate(
      topicKey = Objects.requireNonNull(topicKey),
      numberOfPartitions = CommonUtils.requirePositiveInt(numberOfPartitions),
      numberOfReplications = CommonUtils.requirePositiveShort(numberOfReplications),
      configs = Objects.requireNonNull(configs)
    )

    protected def doCreate(
      topicKey: TopicKey,
      numberOfPartitions: Int,
      numberOfReplications: Short,
      configs: Map[String, String]
    ): Future[Unit]
  }
}

File: ohara-kafka/src/main/java/com/island/ohara/kafka/TopicAdmin.java
/*
 * Copyright 2019 is-land
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.island.ohara.kafka;

import com.island.ohara.common.data.Pair;
import com.island.ohara.common.setting.TopicKey;
import com.island.ohara.common.util.CommonUtils;
import com.island.ohara.common.util.Releasable;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.Properties;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.stream.Collectors;
import org.apache.kafka.clients.CommonClientConfigs;
import org.apache.kafka.clients.admin.AdminClient;
import org.apache.kafka.clients.admin.NewPartitions;
import org.apache.kafka.clients.admin.NewTopic;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.Node;
import org.apache.kafka.common.TopicPartition;
import org.apache.kafka.common.config.ConfigResource;
import org.apache.kafka.common.serialization.ByteArrayDeserializer;

/**
 * a helper methods used by configurator. It provide many helper method to operate kafka cluster.
 * TODO: Do we really need a jave version of "broker" APIs? I don't think it is ok to allow user to
 * touch topic directly... by chia
 *
 * <p>Every method with remote call need to overload in Default Timeout
 */
public interface TopicAdmin extends Releasable {

  /**
   * start a progress to create topic
   *
   * @return creator
   */
  TopicCreator topicCreator();

  /**
   * @param name topic name
   * @return true if topic exists. Otherwise, false
   */
  CompletionStage<Boolean> exist(String name);

  /**
   * @param key topic key
   * @return true if topic exists. Otherwise, false
   */
  default CompletionStage<Boolean> exist(TopicKey key) {
    return exist(key.topicNameOnKafka());
  }

  /**
   * describe the topic existing in kafka. If the topic doesn't exist, exception will be thrown
   *
   * @param key topic key
   * @return TopicDescription
   */
  default CompletionStage<TopicDescription> topicDescription(TopicKey key) {
    return topicDescription(key.topicNameOnKafka());
  }

  default CompletionStage<TopicDescription> topicDescription(String name) {
    return topicDescriptions(Collections.singleton(name))
        .thenApply(
            tds -> {
              if (tds.isEmpty()) throw new NoSuchElementException(name + " does not exist");
              else return tds.get(0);
            });
  }

  default CompletionStage<List<TopicDescription>> topicDescriptions(Set<String> topicNames) {
    return topicDescriptions()
        .thenApply(
            tds ->
                tds.stream()
                    .filter(td -> topicNames.stream().anyMatch(n -> n.equals(td.name())))
                    .collect(Collectors.toList()));
  }

  /**
   * list all topics details from kafka
   *
   * @return topic details
   */
  CompletionStage<List<TopicDescription>> topicDescriptions();

  /**
   * create new partitions for specified topic
   *
   * @param name topic name
   * @param numberOfPartitions number of new partitions
   * @return a async callback is tracing the result
   */
  CompletionStage<Void> createPartitions(String name, int numberOfPartitions);

  /**
   * create new partitions for specified topic
   *
   * @param key topic key
   * @param numberOfPartitions number of new partitions
   * @return a async callback is tracing the result
   */
  default CompletionStage<Void> createPartitions(TopicKey key, int numberOfPartitions) {
    return createPartitions(key.topicNameOnKafka(), numberOfPartitions);
  }

  /**
   * remove topic.
   *
   * @param key topic key
   * @return a async callback is tracing the result. It carries the "true" if it does remove a
   *     topic. otherwise, false
   */
  default CompletionStage<Boolean> deleteTopic(TopicKey key) {
    return deleteTopic(key.topicNameOnKafka());
  }

  /**
   * remove topic.
   *
   * @param name topic encoded name
   * @return a async callback is tracing the result. It carries the "true" if it does remove a
   *     topic. otherwise, false
   */
  CompletionStage<Boolean> deleteTopic(String name);

  /** @return Connection information. form: host:port,host:port */
  String connectionProps();

  /**
   * list all active brokers' ports. This is different to {@link #connectionProps()} since this
   * method will fetch the "really" active brokers from cluster. {@link #connectionProps()} may
   * include dead or nonexistent broker nodes.
   *
   * @return active brokers' ports
   */
  CompletionStage<Map<String, Integer>> brokerPorts();

  boolean closed();

  static TopicAdmin of(String connectionProps) {
    return new TopicAdmin() {

      private final AdminClient admin = AdminClient.create(toAdminProps(connectionProps));

      private final AtomicBoolean closed = new AtomicBoolean(false);

      private Map<String, List<PartitionInfo>> partitionInfos() {
        Properties properties = new Properties();
        properties.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, connectionProps);
        properties.setProperty(ConsumerConfig.CLIENT_ID_CONFIG, CommonUtils.randomString());
        try (KafkaConsumer<byte[], byte[]> consumer =
            new KafkaConsumer<>(
                properties, new ByteArrayDeserializer(), new ByteArrayDeserializer())) {
          return consumer.listTopics().entrySet().stream()
              .map(
                  entry -> {
                    List<TopicPartition> tps =
                        entry.getValue().stream()
                            .map(p -> new TopicPartition(p.topic(), p.partition()))
                            .collect(Collectors.toList());
                    Map<TopicPartition, Long> beginningOffsets = consumer.beginningOffsets(tps);
                    Map<TopicPartition, Long> endOffsets = consumer.endOffsets(tps);
                    List<PartitionInfo> ps =
                        entry.getValue().stream()
                            .map(
                                p -> {
                                  TopicPartition tp = new TopicPartition(p.topic(), p.partition());
                                  return new PartitionInfo(
                                      p.partition(),
                                      PartitionNode.of(p.leader()),
                                      p.replicas() == null
                                          ? Collections.emptyList()
                                          : Arrays.stream(p.replicas())
                                              .map(PartitionNode::of)
                                              .collect(Collectors.toList()),
                                      p.inSyncReplicas() == null
                                          ? Collections.emptyList()
                                          : Arrays.stream(p.inSyncReplicas())
                                              .map(PartitionNode::of)
                                              .collect(Collectors.toList()),
                                      beginningOffsets.getOrDefault(tp, -1L),
                                      endOffsets.getOrDefault(tp, -1L));
                                })
                            .collect(Collectors.toList());
                    return Pair.of(entry.getKey(), ps);
                  })
              .collect(Collectors.toMap(Pair::left, Pair::right));
        }
      }

      @Override
      public TopicCreator topicCreator() {
        return new TopicCreator() {
          @Override
          protected CompletionStage<Void> doCreate(
              int numberOfPartitions,
              short numberOfReplications,
              Map<String, String> options,
              String name) {
            CompletableFuture<Void> f = new CompletableFuture<>();
            admin
                .createTopics(
                    Collections.singletonList(
                        new NewTopic(name, numberOfPartitions, numberOfReplications)
                            .configs(options)))
                .values()
                .get(name)
                .whenComplete(
                    (v, exception) -> {
                      if (exception != null) f.completeExceptionally(exception);
                      else f.complete(null);
                    });
            return f;
          }
        };
      }

      @Override
      public CompletionStage<Boolean> exist(String name) {
        return topicNames().thenApply(tps -> tps.contains(name));
      }

      private CompletionStage<Set<String>> topicNames() {
        CompletableFuture<Set<String>> f = new CompletableFuture<>();
        admin
            .listTopics()
            .names()
            .whenComplete(
                (tps, exception) -> {
                  if (exception != null) f.completeExceptionally(exception);
                  else f.complete(tps);
                });
        return f;
      }

      private CompletionStage<Map<String, List<TopicOption>>> options(Set<String> topicNames) {
        return topicNames()
            .thenApply(
                names ->
                    names.stream()
                        .map(name -> new ConfigResource(ConfigResource.Type.TOPIC, name))
                        .collect(Collectors.toList()))
            .thenCompose(
                rs -> {
                  CompletableFuture<Map<String, List<TopicOption>>> f = new CompletableFuture<>();
                  admin
                      .describeConfigs(rs)
                      .all()
                      .whenComplete(
                          (configs, exception) -> {
                            if (exception != null) f.completeExceptionally(exception);
                            else
                              f.complete(
                                  configs.entrySet().stream()
                                      .map(
                                          entry ->
                                              Pair.of(
                                                  entry.getKey().name(),
                                                  entry.getValue().entries().stream()
                                                      .map(
                                                          o ->
                                                              new TopicOption(
                                                                  o.name(),
                                                                  o.value(),
                                                                  o.isDefault(),
                                                                  o.isSensitive(),
                                                                  o.isReadOnly()))
                                                      .collect(Collectors.toList())))
                                      .collect(Collectors.toMap(Pair::left, Pair::right)));
                          });
                  return f;
                });
      }

      @Override
      public CompletionStage<List<TopicDescription>> topicDescriptions() {
        return topicNames()
            .thenCompose(this::options)
            .thenApply(
                nameAndOpts -> {
                  Map<String, List<PartitionInfo>> partitionInfos = partitionInfos();
                  return nameAndOpts.entrySet().stream()
                      .map(
                          entry -> {
                            List<PartitionInfo> infos =
                                partitionInfos.getOrDefault(
                                    entry.getKey(), Collections.emptyList());
                            return new TopicDescription(entry.getKey(), infos, entry.getValue());
                          })
                      .collect(Collectors.toList());
                });
      }

      @Override
      public CompletionStage<Void> createPartitions(String name, int numberOfPartitions) {
        return topicDescription(name)
            .thenCompose(
                current -> {
                  if (current.numberOfPartitions() > numberOfPartitions)
                    throw new IllegalArgumentException(
                        "Reducing the number from partitions is disallowed. current:"
                            + current.numberOfPartitions()
                            + ", expected:"
                            + numberOfPartitions);
                  else if (current.numberOfPartitions() == numberOfPartitions)
                    return CompletableFuture.completedFuture(null);
                  else {
                    CompletableFuture<Void> f = new CompletableFuture<>();
                    admin
                        .createPartitions(
                            Collections.singletonMap(
                                name, NewPartitions.increaseTo(numberOfPartitions)))
                        .values()
                        .get(name)
                        .whenComplete(
                            (v, exception) -> {
                              if (exception != null) f.completeExceptionally(exception);
                              else f.complete(null);
                            });
                    return f;
                  }
                });
      }

      @Override
      public CompletionStage<Boolean> deleteTopic(String topicName) {
        CompletableFuture<Boolean> f = new CompletableFuture<>();
        topicNames()
            .thenApply(names -> names.stream().anyMatch(name -> name.equals(topicName)))
            .whenComplete(
                (existent, exception) -> {
                  if (exception != null) f.completeExceptionally(exception);
                  else f.complete(existent);
                });
        return f.thenCompose(
            existent -> {
              if (existent) {
                CompletableFuture<Boolean> f2 = new CompletableFuture<>();
                admin
                    .deleteTopics(Collections.singletonList(topicName))
                    .values()
                    .get(topicName)
                    .whenComplete(
                        (v, exception) -> {
                          if (exception != null) f2.completeExceptionally(exception);
                          else f2.complete(true);
                        });
                return f2;
              } else return CompletableFuture.completedFuture(false);
            });
      }

      @Override
      public String connectionProps() {
        return connectionProps;
      }

      @Override
      public CompletionStage<Map<String, Integer>> brokerPorts() {
        CompletableFuture<Map<String, Integer>> f = new CompletableFuture<>();
        admin
            .describeCluster()
            .nodes()
            .whenComplete(
                (nodes, exception) -> {
                  if (exception != null) f.completeExceptionally(exception);
                  else f.complete(nodes.stream().collect(Collectors.toMap(Node::host, Node::port)));
                });
        return f;
      }

      /**
       * this impl will host a kafka.AdminClient so you must call the #close() to release the
       * kafka.AdminClient.
       *
       * @param brokers the kafka brokers information
       * @return the props for TopicAdmin
       */
      private Properties toAdminProps(String brokers) {
        Properties adminProps = new Properties();
        adminProps.put(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, brokers);
        return adminProps;
      }

      @Override
      public void close() {
        if (closed.compareAndSet(false, true)) {
          admin.close();
        }
      }

      @Override
      public boolean closed() {
        return closed.get();
      }
    };
  }
}
